pkg/blob/controllerserver.go

```diff
@@ -667,6 +689,107 @@ func (d *Driver) DeleteBlobContainer(ctx context.Context, subsID, resourceGroupN
+	}
+	if dstContainerName == "" {
+		return fmt.Errorf("dstContainerName is empty")
+	}
+	if len(secrets) > 0 {
+		accountSasToken, err := getAccountSasToken(secrets, accountName)
+		if err != nil || accountSasToken == "" {
+			klog.V(2).Infof("get account sas token from secrets err: %s, generate sas token for account(%s)", err.Error(), accountName)
+			accountSasToken, err = generateSASToken(accountName, accountKey, storageEndpointSuffix)
+			if err != nil {
+				return err
+			}
+		}
+		srcContainer, getErr := getContainerReference(srcContainerName, secrets, d.cloud.Environment)
+		if getErr != nil {
+			return getErr
+		}
+		exists, err := srcContainer.Exists()
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("source container(%s) does not exist", srcContainerName)
+		}
+
+		if err := d.waitForAzCopySuccess(ctx, accountName, resourceGroupName, storageEndpointSuffix, srcContainerName, dstContainerName, accountSasToken, waitForCopyInterval, waitForCopyTimeout); err != nil {
+			return fmt.Errorf("failed to wait for azcopy success: %v", err.Error())
+		}
+	} else {
+		klog.V(2).Infof("generate sas token for account(%s)", accountName)
+		accountSasToken, genErr := generateSASToken(accountName, accountKey, storageEndpointSuffix)
+		if genErr != nil {
+			return genErr
+		}
+		if _, rerr := d.cloud.BlobClient.GetContainer(ctx, subsID, resourceGroupName, accountName, srcContainerName); rerr != nil {
+			return fmt.Errorf("srcContainer(%s) get error: %s", srcContainerName, rerr.Error())
+		}
+
+		if err := d.waitForAzCopySuccess(ctx, accountName, resourceGroupName, storageEndpointSuffix, srcContainerName, dstContainerName, accountSasToken, waitForCopyInterval, waitForCopyTimeout); err != nil {
+			return fmt.Errorf("failed to wait for azcopy success: %v", err.Error())
+		}
+	}
+	return nil
+}
+
+// waitForAzCopySuccess wait for azcopy blob container success
+func (d *Driver) waitForAzCopySuccess(ctx context.Context, accountName, resourceGroupName, storageEndpointSuffix, srcContainerName, dstContainerName, accountSasToken string, intervel, timeout time.Duration) error {
+	timeAfter := time.After(timeout)
+	timeTick := time.Tick(intervel)
+	srcPath := fmt.Sprintf("https://%s.blob.%s/%s%s", accountName, storageEndpointSuffix, srcContainerName, accountSasToken)
+	dstPath := fmt.Sprintf("https://%s.blob.%s/%s%s", accountName, storageEndpointSuffix, dstContainerName, accountSasToken)
+
+	for {
+		select {
+		case <-timeTick:
+			klog.V(2).Infof("copy blob container %s to %s", srcContainerName, dstContainerName)
+			out, copyErr := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false").CombinedOutput()
+			if copyErr != nil {
+				if strings.Contains(string(out), azcopyAuthenticationFailedCode) {
+					klog.Warningf("CopyContainer(%s, %s, %s) failed with authentication error(%v): %v, retrying", resourceGroupName, accountName, dstContainerName, copyErr, string(out))
```

Comment:
what's the difference between `azcopyAuthenticationFailedCode` and non `azcopyAuthenticationFailedCode` error? should we return error no matter it's azcopyAuthenticationFailedCode error?

Reply:
It has AuthenticationFailed error in before and and the reason is for accountKey nil. I will remove this.